ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json

```diff
@@ -572,6 +572,143 @@
         ]
       }
     },
+    {
+      "name": "multiple expressions - KAFKA to struct key format",
+      "statements": [
+        "CREATE STREAM TEST (f1 INT KEY, f2 INT) WITH (kafka_topic='test_topic', value_format='JSON', partitions=1);",
+        "CREATE TABLE OUTPUT WITH (KEY_FORMAT='JSON') AS SELECT STRUCT(f1:=f1, f2:=f2) AS k, as_value(struct(f1:=f1, f2:=f2)) AS key, COUNT(*) AS total FROM TEST GROUP BY STRUCT(f1:=f1, f2:=f2);"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": 1, "value": {"f2": "2"}},
+        {"topic": "test_topic", "key": 2, "value": {"f2": "4"}},
+        {"topic": "test_topic", "key": 1, "value": {"f2": null}},
+        {"topic": "test_topic", "key": 1, "value": null},
+        {"topic": "test_topic", "key": null, "value": {"f2": "1"}},
+        {"topic": "test_topic", "key": 1, "value": {"f2": "2"}},
+        {"topic": "test_topic", "key": 2, "value": {"f2": "4"}},
+        {"topic": "test_topic", "key": 2, "value": {"f2": "1"}}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": {"F1": 1, "F2": 2}, "value": {"KEY": {"F1":1,"F2":2}, "TOTAL": 1}},
+        {"topic": "OUTPUT", "key": {"F1": 2, "F2": 4}, "value": {"KEY": {"F1":2,"F2":4}, "TOTAL": 1}},
+        {"topic": "OUTPUT", "key": {"F1": 1, "F2": null}, "value": {"KEY": {"F1":1,"F2":null}, "TOTAL": 1}},
+        {"topic": "OUTPUT", "key": {"F1": null, "F2": 1}, "value": {"KEY": {"F1":null,"F2":1}, "TOTAL": 1}},
+        {"topic": "OUTPUT", "key": {"F1": 1, "F2": 2}, "value": {"KEY": {"F1":1,"F2":2}, "TOTAL": 2}},
+        {"topic": "OUTPUT", "key": {"F1": 2, "F2": 4}, "value": {"KEY": {"F1":2,"F2":4}, "TOTAL": 2}},
+        {"topic": "OUTPUT", "key": {"F1": 2, "F2": 1}, "value": {"KEY": {"F1":2,"F2":1}, "TOTAL": 1}}
+      ]
+    },
+    {
+      "name": "multiple expressions - KAFKA to non-supported decimal key format",
+      "statements": [
+        "CREATE STREAM TEST (f1 INT KEY, f2 INT) WITH (kafka_topic='test_topic', value_format='JSON', partitions=1);",
+        "CREATE TABLE OUTPUT WITH (KEY_FORMAT='JSON') AS SELECT ROUND(f1) AS key, COUNT(*) AS total FROM TEST GROUP BY ROUND(f1);"
```

Comment:
The output schema of `ROUND(int)` is a long, not a decimal, so this test does not test what it claims to test.

You can add a post-test assertion to assert the expected schema: https://github.com/confluentinc/ksql/blob/3bf8531d0f67de2590ed8f945033b58f68e38a0f/ksqldb-functional-tests/src/test/resources/query-validation-tests/group-by.json#L28

Reply:
I see, how do we create a decimal type then?

Reply:
You can create a decimal-type column in the source stream and GROUP BY that column: https://docs.ksqldb.io/en/latest/reference/sql/data-types/#arbitrary-precision-types